TICL version 4: CLUE3D as default Pattern Recognition, Linking

waredjeb opened the pull request:
This PR introduces TICLv4:
- Using CLUE3D as default pattern recognition. Replaced the 4 Iterations with a single global one.
- Enabling the possibility to fall back to TICLv3 with `--procModifiers ticl_v3`.
- Tackster-Trackster and Trackster-Track geometrical linking. Propagating Tracksters and Tracks to
  the same plane, if they fall in the same eta-phi window they are geometrically compatible. Trackster
  and Tracks are then checked for Time compatibility. They are time compatible if within 3sigma of each other
  or either of them does not have time available. Possible charged candidates are checked for energy compatibility
  with the corresponding track. TICLCandidate and Track are energy compatible if the TICLCandidate's tracksters
  cumulative energy does not exceed track momentum by more than a threshold. Track with no matched Trackster are
  promoted to charged-TICLCandidate. Note that muon tracks are masked and not used for the linking. 
- Improved Particle Flow interpretation

- More details on the Linking step here: [Linking](https://indico.cern.ch/event/1167627/contributions/4906008/attachments/2454845/4207362/ticlv4-linking.pdf)


-  CLUE3D performance compared to Cellular Automaton: 

    - [CLUE3DvsCA](https://indico.cern.ch/event/1160548/contributions/4873846/attachments/2443268/4186202/Updates%20on%20energy%20reconstruction%20with%20PU%20(1).pdf)  
    - [CLUE3DvsCATiming](https://indico.cern.ch/event/1162565/contributions/4885086/attachments/2447097/4193233/Updates%20on%20Clue3D%20performance%20with%20PU%20(3).pdf)

PR validated on CMSSW_12_5_0_pre3. 
- Timing comparison between TICLv4 and TICLv3, obtained on RelVal QCD PU 200, 10 events: 
  - [TICLv4](https://wredjeb.web.cern.ch/HGCAL/TICLv4PFValidation/web/piechart.php?local=false&dataset=resources&resource=time_real&colours=default&groups=packages&threshold=0)
  - [TICLv3](https://wredjeb.web.cern.ch/HGCAL/TICLv4PFValidation/web/piechart.php?local=false&dataset=resources_v3&resource=time_real&colours=default&groups=packages&threshold=0)
  
 Physics Performances: 
 - RecoParticleFlow Validation on `RelValQCD_Pt15To7000_Flat_14` PU200, D88 Geometry:
   - [PFValidation](https://wredjeb.web.cern.ch/HGCAL/TICLv4PFValidation/plots_v3_v4_pu200/)

**EDIT 08/08 :** 
- Small bug fix, setting Trackster `seedIndex` default value to -1,  since 0 is a valid `seedIndex`.

felicepantaleo committed:
TICL v4 version, CLUE3D as default Pattern Recognition, Trackster/Track Linking
This commit is a set of multiple commits:
- Using CLUE3D as default pattern recognition. Replaced the 4 Iterations with a single global one.
  Enabling possibility to fall back to TICLv3 with `--procModifier ticl_v3`.
- Tackster-Trackster and Trackster-Track geometrical linking. Propagating Tracksters and Tracks to
  the same plane, if they fall in the same eta-phi window they are geometrically compatible. Trackster
  and Tracks are then checked for Time compatibility. They are time compatible if within 3sigma of each other
  or either of them does not have time available. Possible charged candidates are checked for energy compatibility
  with the corresponding track. TICLCandidate and Track are energy compatible if the TICLCandidate's tracksters
  cumulative energy does not exceed track momentum by more than a threshold. Track with no matched Trackster are
  promoted at charged-TICLCandidate. Note that muon tracks are masked and not used for the linking.
   Improving Particle Flow interpreation.

Co-Authored-By: Abhirikshma Nandi <abhirikshma14@gmail.com>
Co-Authored-By: Marco Rovere <marco.rovere@cern.ch>
Co-Authored-By: Felice Pantaleo <felice.pantaleo@cern.ch>

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31308

ERROR: Build errors found during clang-tidy run.
```
RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:91:12: error: variable-sized object may not be initialized [clang-diagnostic-error]
  int mask[trackstersSize] = {0};
           ^
RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:329:19: error: variable-sized object may not be initialized [clang-diagnostic-error]
  int chargedMask[tracksters.size()] = {0};
                  ^
RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:458:9: warning: the 'empty' method should be used to check for emptiness instead of 'size' [readability-container-size-empty]
--
RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:476:19: error: variable-sized object may not be initialized [clang-diagnostic-error]
  int neutralMask[tracksters.size()] = {0};
                  ^
RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:519:11: warning: the 'empty' method should be used to check for emptiness instead of 'size' [readability-container-size-empty]
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31311

- This PR adds an extra 68KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #38760, #38834, #38702, #33532

cmsbuild commented:
A new Pull Request was created by  @waredjeb (Wahid Redjeb) for master.

It involves the following packages:

- Configuration/ProcessModifiers (**operations**)
- Configuration/PyReleaseValidation (**pdmv, upgrade**)
- RecoHGCal/TICL (**upgrade, reconstruction**)
- Validation/HGCalValidation (**dqm**)


@perrotta, @rappoccio, @jordan-martins, @bbilin, @pmandrik, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @kskovpen, @jpata, @qliphy, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@felicepantaleo, @forthommel, @kpedro88, @Martin-Grunewald, @bsunanda, @pfs, @trtomei, @slomeo, @youyingli, @sethzenz, @apsallid, @makortel, @lgray, @missirol, @beaucero, @vandreev11, @rovere, @cseez, @hatakeyamak, @ebrondol, @fabiocos, @sobhatta, @lecriste this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

rovere commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26506/summary.html
**COMMIT**: ccabda0476119cfb6eb0e6b3ec902048f3405865
**CMSSW**: CMSSW_12_5_X_2022-07-28-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/26506/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [39434.75](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26506/runTheMatrix-results/39434.75_TTbar_14TeV+2026D88_HLT75e33+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HLT75e33+HARVESTGlobal)
```
----- Begin Fatal Exception 28-Jul-2022 16:15:41 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=TrackstersMergeProducer label='ticlTrackstersMerge'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'seedingTrk'
 'trackstersem'
 'trackstershad'
 'tracksterstrk'
 'tracksterstrkem'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+    else if (seed_eta < 0.) {
+      // backward tile
+      const TICLLayerTile &tile = tracksterTiles[0];
+      double eta_min = std::max(abs(seed_eta) - delta, (double)TileConstants::minEta);
+      double eta_max = std::min(abs(seed_eta) + delta, (double)TileConstants::maxEta);
+
+      std::array<int, 4> search_box = tile.searchBoxEtaPhi(eta_min, eta_max, seed_phi - delta, seed_phi + delta);
+      if (search_box[2] > search_box[3])
+        search_box[3] += TileConstants::nPhiBins;
+
+      for (int eta_i = search_box[0]; eta_i <= search_box[1]; ++eta_i) {
+        for (int phi_i = search_box[2]; phi_i <= search_box[3]; ++phi_i) {
+          const auto &in_box = tile[tile.globalBin(eta_i, (phi_i % TileConstants::nPhiBins))];
+          for (const unsigned t_i : in_box) {
+            if (!mask[t_i]) {
+              resultCollection[seedId].push_back(t_i);
+              if (useMask)
+                mask[t_i] = 1;
+            }
+          }
+        }
+      }
```
this seems identical to the branch above. any chance you can define it as a common function to avoid code duplication?

waredjeb replied:
thanks for the comments. Sure, I can avoid code duplication

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+
+void LinkingAlgoByDirectionGeometric::dumpLinksFound(std::vector<std::vector<unsigned>> &resultCollection,
+                                                     const char *label) const {
+  if (!(LinkingAlgoBase::algo_verbosity_ > LinkingAlgoBase::Advanced))
```
you could also wrap the contents of this function in `#ifdef EDM_ML_DEBUG` so that it's compiled if and only if debug is enabled.
it may be confusing for non-TICL people in the future to navigate multiple parallel levels of message logging.

jpata replied:
now the function implementation is not defined if `EDM_ML_DEBUG` is not defined. In fact I'm confused why scram does not crash at linking due to a missing implementation.

I thought something like this might be desirable here:
```C++
void LinkingAlgoByDirectionGeometric::dumpLinksFound(std::vector<std::vector<unsigned>> &resultCollection,
                                                     const char *label) const {
#ifdef EDM_ML_DEBUG
...
#endif
}
```

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+                                                     const std::vector<reco::Muon> &muons,
+                                                     const edm::Handle<std::vector<Trackster>> tsH,
+                                                     std::vector<TICLCandidate> &resultLinked) {
+  constexpr double mpion = 0.13957;
```
it might be good to define the TICL pi mass in a constexpr somewhere, also replacing

```
RecoHGCal/TICL/plugins/TICLCandidateFromTrackstersProducer.cc:          constexpr double mpion2 = 0.13957 * 0.13957;
RecoHGCal/TICL/plugins/TracksterP4FromTrackAndPCA.cc:        constexpr double mpion2 = 0.13957 * 0.13957;
RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:  constexpr double mpion = 0.13957;
```

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+    reco::TrackRef trackref = reco::TrackRef(tkH, i);
+    // also veto tracks associated to muons
+    int muId = PFMuonAlgo::muAssocToTrack(trackref, muons);
+    if (LinkingAlgoBase::algo_verbosity_ > LinkingAlgoBase::Advanced)
```
it seems strange to protect a compile-time-enabled message with a runtime flag.

waredjeb replied:
Hi! We use this strategy to have different log-debug levels in TICL, but if this is a problem I can remove it

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+    candidateTrackIds.push_back(i);
+
+    // don't consider tracks below 2 GeV for linking
+    if (std::sqrt(tk.p() * tk.p() + mpion2) < 2.0)
```
it might useful if this 2 GeV cut was a constexpr somewhere in the class configuration.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+          << t.time() << " energy " << t.raw_energy() << "\n";
+    Vector directnv = t.eigenvectors(0);
+
+    if (abs(directnv.Z()) < 0.00001)
```
this magic number should be a constexpr.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+      LogDebug("LinkingAlgoByDirectionGeometric")
+          << "trackster " << i << " - eta " << t.barycenter().eta() << " phi " << t.barycenter().phi() << " time "
+          << t.time() << " energy " << t.raw_energy() << "\n";
+    Vector directnv = t.eigenvectors(0);
```
```suggestion
    const auto& directnv = t.eigenvectors(0);
```
perhaps?

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+
+    // to HGCal front
+    float zVal = hgcons_->waferZ(1, true);
+    Vector tsP = propagateTrackster(t, i, zVal, tracksterPropTiles);
```
likewise here - do you need to copy the vector?

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+    auto energyCompatible = [&](const Trackster &ts, const reco::Track tk) -> bool {
+      // compatible if accumulated energy does not
+      // exceed track momentum by more than threshold
+      double threshold = std::min(0.2 * ts.raw_energy(), 10.0);
```
magic numbers as constexpr.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+      double tkT = tkTime[tk];
+      double tkTErr = tkTimeErr[tk];
+
+      if (tsT == -99. or tkTimeQual[tk] < timing_quality_threshold_)
```
is -99.f defined as a magic number for an uninitialized trackster time? if so, it would be good if it was defined as a constexpr to prevent confusion in the future.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,611 @@
+    auto tkRef = reco::TrackRef(tkH, i);
+
+    for (const unsigned ts3_idx : tsNearTk[i]) {  // tk -> ts
+      if (!chargedMask[ts3_idx]) {
+        if (!energyCompatible(tracksters[ts3_idx], tracks[i]) or !timeCompatible(tracksters[ts3_idx], tkRef))
+          continue;
+        chargedCandidate.addTrackster(edm::Ptr<Trackster>(tsH, ts3_idx));
+        chargedMask[ts3_idx] = 1;
+        total_raw_energy += tracksters[ts3_idx].raw_energy();
```
perhaps consider a function to reduce the code repetition in the next 8 blocks.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.h:
```diff
@@ -0,0 +1,92 @@
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/CommonDetUnit/interface/GeomDet.h"
+
+//#include "RecoHGCal/TICL/test/PCA_mod.h"
```
remove

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -197,348 +265,118 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
-    printTrackstersDebug(trackstersTRK, "tracksterTRK");
-    printTrackstersDebug(trackstersHAD, "tracksterHAD");
-  }
+  edm::Handle<std::vector<reco::Muon>> muons_h;
+  evt.getByToken(muons_token_, muons_h);
+  const auto &muons = *muons_h;
```
```suggestion
  const auto& muons = evt.get(muons_token_);
```
also below

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -197,348 +265,118 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
-    tmpCandidate.setP4(p4);
-    resultCandidates->push_back(tmpCandidate);
+  // Print debug info
+  if (debug_) {
```
same comment as above about the necessity of additional runtime flags in front of compile-time-enabled logging.

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc:
```diff
@@ -0,0 +1,776 @@
+  }
+
+  // Neutral Hadrons
+  constexpr double mpion = 0.13957;
```
constexpr.

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc:
```diff
@@ -0,0 +1,776 @@
+            tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, mergedIdx));
+            double raw_energy = t.raw_energy();
+
+            tmpCandidate.setCharge(track.charge());
+            tmpCandidate.setTrackPtr(edm::Ptr<reco::Track>(track_h, trackIdx));
+            tmpCandidate.setPdgId(211 * track.charge());
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+              raw_energy += trackstersMergedHandle->at(otherTracksterIdx).raw_energy();
+            }
+            tmpCandidate.setRawEnergy(raw_energy);
+            math::XYZTLorentzVector p4(raw_energy * track.momentum().unit().x(),
+                                       raw_energy * track.momentum().unit().y(),
+                                       raw_energy * track.momentum().unit().z(),
+                                       raw_energy);
+            tmpCandidate.setP4(p4);
+            resultCandidates->push_back(tmpCandidate);
```
consider reducing the repetition with respect to the code below by using a function

waredjeb replied:
I think in this case a function would make the code readability harder, I will rearrange the code a bit to avoid duplication as much as possible, but I think it would be better to have all the logic to build the candidates more explicit.

jpata replied:
```C++
if (foundCompatibleTRK) {
  resultCandidates->push_back(buildChargedHadron(...));
}
} else {
  resultCandidates->push_back(buildElectron(...));
}
...
```
I would suggest something like this here and below. Having code blocks that are hundreds of lines long seems difficult to read and maintain later.

Abhirikshma replied:
Hello @jpata, thanks for the comments! I think implementing functions is slightly unnecessary here:
- this is part of the TICL v3 logic which is deprecated and will be removed soon (and indeed this code is already in production as `TrackstersMergeProducer.cc`, we just changed the name of the file to keep it there till everything switches to using TICL v4)
- It would require some non-trivial modification of the code, and we risk introducing bugs in something which is already very well tested

jpata reviewed (commented):
Some general code comments:
- avoid hardcoding magic numbers, this makes the code hard to maintain in the future
- try to reduce copy-paste parts if possible (e.g. when making particle candidates)
- it might be good to use a single logging configuration across CMS. Do you really need to introduce another runtime logging level on top of the EDM MessageLogger?
- vectors are often not preallocated. did you confirm that it's never possible?

Can you also provide some representative profiling (CPU and memory) information of the new algo?

waredjeb committed:
avoid code duplication, rearrange code

waredjeb committed:
use TracksterMergeV3 for HLT75e33 wfs

waredjeb committed:
set trackster seedIndex to -1

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31459

- This PR adds an extra 108KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #38760, #38702, #38967, #33532

cmsbuild commented:
Pull request #38880 was updated. @perrotta, @rappoccio, @jordan-martins, @missirol, @bbilin, @pmandrik, @emanueleusai, @ahmad3213, @Martin-Grunewald, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @kskovpen, @jpata, @qliphy, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26703/summary.html
**COMMIT**: 4b296b541c82e63502855507372ef976aace1922
**CMSSW**: CMSSW_12_5_X_2022-08-08-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/26703/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 96908 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 10293 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3590852
 - DQMHistoTests: Total failures: 22604
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3568225
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -80769.819 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): -10466.235 KiB   HGCAL/HGCalValidator
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - DQMHistoSizes: changed ( 39434.0,... ): -9874.222 KiB    HGCAL/HGCalValidator
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-08-1100+51a7ed/52028/triggerResults): no differences found

kpedro88 commented:
> You potentially added 96908 lines to the logs

This should be fixed...

kpedro88 commented:
Appears to be the result of an actual bug; this message is repeated many times in Phase 2 step 3 workflow logs:
```
%MSG-w PFClusterAlgo:  PFClusterProducer:particleFlowClusterHGCal  08-Aug-2022 16:07:53 CEST Run: 1 Event: 1
rechit 2317077666 has a NaN energy... The input of the particle flow clustering seems to be corrupted.
%MSG
```

waredjeb committed:
bug fix in recordTrackster function

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31482

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #38760, #38702, #38967, #33532

cmsbuild commented:
Pull request #38880 was updated. @perrotta, @Martin-Grunewald, @rappoccio, @jordan-martins, @missirol, @bbilin, @pmandrik, @emanueleusai, @ahmad3213, @cmsbuild, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @kskovpen, @jpata, @qliphy, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26738/summary.html
**COMMIT**: cb53fdb9b639564c5175bc8c11212ed74ed677ef
**CMSSW**: CMSSW_12_5_X_2022-08-10-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/26738/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 9987 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3591716
 - DQMHistoTests: Total failures: 22300
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3569393
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -80769.811 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): -10466.235 KiB   HGCAL/HGCalValidator
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - DQMHistoSizes: changed ( 39434.0,... ): -9874.222 KiB    HGCAL/HGCalValidator
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-10-1100+51a7ed/52061/triggerResults): no differences found

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -197,348 +265,107 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+  for (const auto &cand : candidates) {
+    auto track_ptr = cand.trackPtr();
+    auto trackster_ptrs = cand.tracksters();
+    if (debug_) {
```
this block could all be in `#ifdef EDM_ML_DEBUG`, as LogDebug is only enabled in that case.

Abhirikshma replied:
thanks, I have removed the runtime `debug_` flag completely and wrapped the actions associated to message logging in `#ifdef EDM_ML_DEBUG`

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -197,348 +265,107 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+    ticl::Trackster outTrackster;
+    auto updated_size = 0;
+    for (const auto &ts_ptr : trackster_ptrs) {
+      if (debug_) {
```
same here

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc:
```diff
@@ -0,0 +1,776 @@
+
+  // associating seed to the index of the trackster in the merged collection and the iteration that found it
+  std::map<int, std::vector<std::pair<int, TracksterIterIndex>>> seedToTracksterAssociator;
+  edm::Handle<std::vector<reco::Track>> track_h;
+  evt.getByToken(tracks_token_, track_h);
+  const auto &tracks = *track_h;
```
```suggestion
  const auto &tracks = evt.get(tracks_token_);
```
here and below. (I think this comment got lost and was not implemented from earlier)

Abhirikshma replied:
this is also done except for the cases where we require the `edm::Handle`s later to build the `TICLCandidates`

jpata commented on RecoHGCal/TICL/plugins/TrackstersMergeProducerV3.cc:
```diff
@@ -0,0 +1,776 @@
+
+  int counter = 0;
+  for (auto const &t : tracksters) {
+    LogDebug("TrackstersMergeProducerV3")
```
LogDebug is only compiled if EDM_ML_DEBUG is defined at compile time, hence runtime checks can be omitted here.

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,557 @@
+      continue;
+
+    int iSide = int(tk.eta() > 0);
+    FreeTrajectoryState fts = trajectoryStateTransform::outerFreeState((tk), bFieldProd);
```
```suggestion
    const auto& fts = trajectoryStateTransform::outerFreeState((tk), bFieldProd);
```
to avoid a copy

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoByDirectionGeometric.cc:
```diff
@@ -0,0 +1,557 @@
+    int iSide = int(tk.eta() > 0);
+    FreeTrajectoryState fts = trajectoryStateTransform::outerFreeState((tk), bFieldProd);
+    // to the HGCal front
+    TrajectoryStateOnSurface tsos = prop.propagate(fts, firstDisk_[iSide]->surface());
```
also here

jpata commented on RecoHGCal/TICL/plugins/LinkingAlgoBase.h:
```diff
@@ -0,0 +1,52 @@
+
+    static void fillPSetDescription(edm::ParameterSetDescription& desc) { desc.add<int>("algo_verbosity", 0); };
+
+    enum VerbosityLevel { None = 0, Basic, Advanced, Expert, Guru };
```
I would rather prefer not to introduce a module-specific logging configuration and just stick with pure MessageLogger levels.

If you keep this, not only you toyda, but every future reader and maintainer will have two separate levels to keep in mind and get right, rather than one.

felicepantaleo replied:
Hi @jpata , thanks for the comment. Eventually it would be nice to have it uniform across CMSSW with Message Logger levels. 
I asked the active developers if this existing code is something they are actively using in their debugging workflows, and this module is still under heavy development. Do you mind if we keep it like this until the development of this module becomes maintenance and then we roll back not only linking but also other modules that are using these message levels?

jpata replied:
Anything that is merged is already maintenance and someone other than the immediate developers may have to step in to work on it.
I doubt consistent logging will be implemented in the future, if it's already too troublesome to do now.

If you decide to keep this, you could consider using the same enum across different TICL modules, so at least it's possible to keep self consistent in TICL in the future?
Now I see redefinitions of this VerbosityLevel in PatternRecognitionAlgoBase, HGCGraph, SeedingRegionAlgoBase.
For modules in this PR that take a bool flag `debug`, you can prefer the enum, so all of them can be configured in the same way.

Abhirikshma replied:
I have replaced the class wise enums with something that is common for all the TICL modules where this is used

jpata reviewed (commented):
Some outstanding comments.
- the code readability could be further improved by splitting up long blocks into functions
- I would still strongly suggest to stick with vanilla edm MessageLogger levels, rather than introducing per-class levels that can easily get inconsistent between classes.

Martin-Grunewald commented on HLTrigger/Configuration/python/HLT_75e33/modules/ticlTrackstersMerge_cfi.py:
```diff
@@ -1,6 +1,6 @@
 import FWCore.ParameterSet.Config as cms
 
-ticlTrackstersMerge = cms.EDProducer("TrackstersMergeProducer",
+ticlTrackstersMerge = cms.EDProducer("TrackstersMergeProducerV3",
```
Does this mean HLT Phase-2 is running an old version of TICL? When will that version be removed? Is there a plan to update HLT to use TICL V4? @trtomei ?

felicepantaleo replied:
Hi @Martin-Grunewald , yes. 
As I said during the HLT upgrade meeting, the plan is to switch HLT to TICLv4, but this might need some work from POGs to adjust the way they are consuming tracksters.

Martin-Grunewald replied:
@felicepantaleo 
OK, thanks!

Abhirikshma committed:
Make algo verbosity levels consistent across TICL

and minor changes removing runtime debug flag for logging,
change how products are consumed from event.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31708

- This PR adds an extra 60KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #38760, #38702, #33532

cmsbuild commented:
Pull request #38880 was updated. @rappoccio, @bbilin, @Martin-Grunewald, @perrotta, @ahmad3213, @cmsbuild, @missirol, @mandrenguyen, @pmandrik, @emanueleusai, @AdrianoDee, @jfernan2, @kskovpen, @jpata, @qliphy, @micsucmed, @fabiocos, @jordan-martins, @clacaputo, @srimanob, @davidlange6, @rvenditti can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26995/summary.html
**COMMIT**: c183b630c8974ccff0a9a45e57bbde66d230cd83
**CMSSW**: CMSSW_12_5_X_2022-08-22-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/26995/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [39434.75](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/26995/runTheMatrix-results/39434.75_TTbar_14TeV+2026D88_HLT75e33+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HLT75e33+HARVESTGlobal)
```
----- Begin Fatal Exception 22-Aug-2022 23:37:47 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=TrackstersMergeProducerV3 label='ticlTrackstersMerge'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'debug'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38880/31722

- This PR adds an extra 136KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #38760, #38702, #33532

cmsbuild commented:
Pull request #38880 was updated. @perrotta, @rappoccio, @jordan-martins, @missirol, @bbilin, @pmandrik, @emanueleusai, @ahmad3213, @mandrenguyen, @Martin-Grunewald, @AdrianoDee, @srimanob, @jfernan2, @clacaputo, @kskovpen, @jpata, @qliphy, @rvenditti, @micsucmed, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/27008/summary.html
**COMMIT**: f3b9ad5150ee388cb5237451fb0edd56601b14d4
**CMSSW**: CMSSW_12_5_X_2022-08-22-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/27008/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 9850 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3592286
 - DQMHistoTests: Total failures: 22262
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3570002
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -80769.815 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): -10466.235 KiB   HGCAL/HGCalValidator
 - DQMHistoSizes: changed ( 39434.0,... ): -9874.222 KiB    HGCAL/HGCalValidator
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-22-2300+51a7ed/52288/triggerResults): no differences found

felicepantaleo commented:
@jpata we were waiting for this PR to submit the update to the pid/er model used by TICL v4. Do you suggest making it now before this is merged so it can be validated altogether? I don't know how to trigger tests with a PR from the cmsdata repository, but I know it is possible.

Martin-Grunewald commented:
+hlt
HLT Phase-2 to stay with TICL V3 for now, migration of HLT Phase-2 to V4 to be done later.
@trtomei

felicepantaleo commented:
Thanks @jpata OK then we will make the PR of the model, and make a PR after this is merged to actually use the model and test things altogether.
So for what concerns this PR we are good to go!

jpata commented:
+reconstruction
- updates to TICL reconstruction (with corresponding changes in Phase2 outputs)

srimanob commented on Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py:
```diff
@@ -507,6 +507,30 @@ def condition(self, fragment, stepList, key, hasHarvest):
+        'HARVESTGlobal'
+    ],
+    suffix = '_ticl_v3',
+    offset = 0.203,
```
You should update also `Configuration/PyReleaseValidation/README.md`

srimanob commented:
test parameters:
- workflow = 39434.203
- relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,ged,machine,premix

srimanob commented:
@cmsbuild please test

This is just to refresh the test, also with your introduced workflow to roll back to TICLv3.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-51a7ed/27111/summary.html
**COMMIT**: f3b9ad5150ee388cb5237451fb0edd56601b14d4
**CMSSW**: CMSSW_12_5_X_2022-08-25-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38880/27111/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-51a7ed/39434.203_TTbar_14TeV+2026D88_ticl_v3+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal

**Summary**:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 9888 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3594948
 - DQMHistoTests: Total failures: 22338
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3572588
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -80769.815 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): -10466.235 KiB   HGCAL/HGCalValidator
 - DQMHistoSizes: changed ( 39434.0,... ): -9874.222 KiB    HGCAL/HGCalValidator
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-25-2300+51a7ed/52384/triggerResults): no differences found

srimanob commented:
+Upgrade

Move TICL to v4, and introduce the workflow to roll back to v3.

kskovpen commented:
+pdmv

rovere commented:
Would it be possible to have the remaining signatures added in the coming days? It would be nice to have this PR integrated asap.

emanueleusai commented:
+1

rappoccio commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
